src/bsoncxx/test/optional.test.cpp

```diff
@@ -0,0 +1,278 @@
+    CHECK(static_checks<immobile>());
+    CHECK(static_checks<not_copyable>());
+    CHECK(static_checks<not_default_constructible>());
+    CHECK(static_checks<allows_moving_explicit_conversion>());
+    CHECK(static_checks<allows_moving_implicit_conversion>());
+    CHECK(check_conversions<int, int>());
+    CHECK(check_conversions<int, double>());
+    CHECK(check_conversions<double, std::string>());
+    CHECK(check_conversions<std::string, double>());
+    CHECK(check_conversions<std::string, const char*>());
+    CHECK(check_conversions<const char*, std::string>());
+    CHECK(check_conversions<int*, std::unique_ptr<int>>());
+    CHECK(check_conversions<int, allows_moving_explicit_conversion>());
+    CHECK(check_conversions<int, allows_moving_implicit_conversion>());
+    optional<int> v(21);
+    CHECK(*v == 21);
+    CHECK_FALSE(v < v);
+}
+
+TEST_CASE("optional: Nontrivial contents") {
+    optional<std::string> str = "abcd1234";
+    CHECK(str == "abcd1234");
+    {
+        auto dup = str;
+        CHECK(dup == str);
+    }
+    CHECK(str == "abcd1234");
+
+    optional<std::unique_ptr<int>> aptr;
+    CHECK(aptr != nullptr);
+    CHECK(aptr == bsoncxx::stdx::nullopt);
+    {
+        auto dup = std::move(aptr);
+        CHECK(aptr == dup);
+    }
+    aptr = bsoncxx::stdx::make_unique<int>(31);
+    CHECK(aptr != nullopt);
+    REQUIRE(aptr != nullptr);
+    CHECK(**aptr == 31);
+    {
+        auto dup = std::move(aptr);
+        CHECK(aptr);
+        CHECK(aptr == nullptr);
+        REQUIRE(dup);
+        REQUIRE(dup != nullptr);
+        CHECK(**dup == 31);
+    }
+    CHECK(aptr == nullptr);
+}
+
+TEST_CASE("Comparisons") {
+    optional<int> a = 21;
+    optional<int> b = 23;
+    optional<int> c;
+    CHECK(a != nullopt);
+    CHECK(a == 21);
+    CHECK(b != nullopt);
+    CHECK(b == 23);
+    CHECK(c == nullopt);
+    // Null compares less-than values:
+    CHECK(c < 42);
+    CHECK(a < b);
+    CHECK(c < a);
+    CHECK(c < b);
+    CHECK(c == nullopt);
+    CHECK(nullopt < a);
+    CHECK(nullopt == c);
+    CHECK(a != b);
+    CHECK(c != a);
+    CHECK(c != b);
+    CHECK(c == c);
+
+    optional<double> dbl = 3.14;
+    CHECK(dbl != a);
+}
+
+struct in_place_convertible {
+    bool constructed_from_in_place = false;
+    in_place_convertible() = default;
+    in_place_convertible(bsoncxx::stdx::in_place_t) : constructed_from_in_place(true) {}
+};
+
+TEST_CASE("optional<T> conversions") {
+    static_assert(!std::is_constructible<optional<bool>, optional<std::string>>{}, "fail");
+
+    optional<std::string> s1(bsoncxx::stdx::in_place);
+    CHECK(s1 == "");
+
+    optional<in_place_convertible> q(bsoncxx::stdx::in_place);
+    REQUIRE(q.has_value());
+    CHECK_FALSE(q->constructed_from_in_place);
+
+    optional<const char*> c_str = "foo";
+    optional<std::string> string = c_str;
+    optional<std::string> string2 = std::move(c_str);
+    CHECK(string == c_str);
+    CHECK(string2 == c_str);
+}
```

Comment:
Suggest adding more exhaustive tests of use the new API. Example: https://gist.github.com/kevinAlbs/5f3ddccb491f5e4c42633589781e1909

Tests may have helped revealed earlier issues (e.g. converting move constructors and `swap`). These tests may help cross-reference with cppreference to determine any missing API.